Make monitor interval customizable

twang2218 opened the pull request:
1 second interval will cause high CPU usage on Linux system, with this
PR, the interval has been increased to 3 seconds, which reduced the CPU
usage in this case, however, more frequent clipboard register will be
missed.

Fixes https://github.com/tinytacoteam/zazu/issues/189

Signed-off-by: Tao Wang <twang2218@gmail.com>

twang2218 committed:
Make the monitor interval customizable

By default the 100 ms interval in the plugin system  will cause high CPU impact at Linux system for this clipboard plugin. With this PR the interval will be customizable for the user, and the default interval has been increased to 3000 ms for Linux system, which reduced the CPU usage in this case, and for the other system the default interval has been increased to 1000 ms, however, more frequent clipboard register will be missed.

Fixes https://github.com/tinytacoteam/zazu/issues/189

Signed-off-by: Tao Wang <twang2218@gmail.com>

avaly commented:
Could this interval be customizable from the `.zazurc.json` file? It should give more control to users.

twang2218 commented:
@avaly That's a good idea, currently I don't think it's possible to customize the plugin block's configuration.

@blainesch WDYT, should we add such feature?

blainesch commented:
We could kind of customize it. We could set up the service like this:

~~~
const realFunc = () => {
  console.log('the real work')
}

module.exports = (pluginContext) => {
  let interval = null
  return (env = {}) => {
    if (!interval) {
      interval = setInterval(realFunc, env.timeout || 3000)
    }
    Promise.resolve('ping')
  }
}
~~~

What do you think? We could also find a way to make this a core feature, but I'm not sure how useful it would be for most cases.

twang2218 commented:
Yeah, I was just thinking about this way, I will update this PR in this approach.

I'm not quite sure we need put this into the core feature either, we can decide it later if there are more request like this.

twang2218 commented:
Updated the PR to enable the customized update interval, and also increased the default interval.

bayleedev commented on readme.md:
```diff
@@ -67,6 +67,17 @@ By defualt we remember the last 50 items in your clipboard. To change this add a
 }
 ~~~
 
+The clipboard monitor update interval is `1` second (except `3` seconds for Linux system) by default, however, user can customize the interval by setting `updateInterval` option in the configuration. The interval has to be no less than `250` ms.
```
Since everything is in `ms` can we make all examples in ms, not second?

bayleedev replied:
I think the default is `100ms` not `250ms` but maybe I'm wrong?

twang2218 replied:
Sure, I will unify the unit to `ms`.

About the minimal interval, the plugin system's default is `100` ms, however, since the performance issue in the clipboard, I increased the minimum interval to `250` ms.

The reason is it's not only Linux has the performance hit, it's on the macOS as well if the clipboard contains image, image comparison in each round will cost about `1000` ms or more on my laptop if it's a photo, so increasing the minimum interval will let CPU has some rest in the between.

bayleedev commented on src/monitor.js:
```diff
@@ -40,14 +40,46 @@ module.exports = (pluginContext) => {
+    .then(() => monitor(env))
+    .then(() => start(env))
+    .catch((error) => {
+      console.log('error', error)
```
I think the wrapping block will put this in the plugin debugger without you needing to catch/console/throw again.

twang2218 replied:
There reason I put a `catch()` here is that I found the plugin system seems not able to print the exception from plugin. During the development, my plugin had exception several times, however, the log didn't show the details of exception:

```json
$ grep 'Script failed' * | more
zazu.log.2017-02-10:{"plugin":"tinytacoteam/zazu-clipboard","block":"Monitor","level":"error","message":"Script failed","timestamp":"2017-02-10T07:18:36.392Z"}
zazu.log.2017-02-10:{"plugin":"tinytacoteam/zazu-clipboard","block":"Monitor","level":"error","message":"Script failed","timestamp":"2017-02-10T07:19:01.413Z"}
zazu.log.2017-02-10:{"query":"","error":{},"plugin":"tinytacoteam/zazu-clipboard","block":"searcher","level":"error","message":"Script failed","timestamp":"2017-02-10T08:23:15.458Z"}
zazu.log.2017-02-10:{"query":"s","error":{},"plugin":"tinytacoteam/zazu-clipboard","block":"searcher","level":"error","message":"Script failed","timestamp":"2017-02-10T08:23:18.504Z"}
```

I'm not sure why it's not print so I put an extra catch/console/throw here, I will investigate this issue.

twang2218 commented:
Updated this PR with unified time unit in the documents, and removed the extra `catch/console/throw` section. I submitted a PR fix for the log issue in `zazu` project instead.

twang2218 commented:
@blainesch , could you review this PR? Thanks.

bayleedev commented on src/monitor.js:
```diff
@@ -40,14 +40,45 @@ module.exports = (pluginContext) => {
+        if (!lastClip || lastClip.type !== clip.type || lastClip.raw !== clip.raw) {
+          lastClip = clip
+          const clipCollection = CappedClient.init(cwd, env)
+          resolve(clipCollection.upsert(clip))
```
Nice

bayleedev commented on src/monitor.js:
```diff
@@ -40,14 +40,45 @@ module.exports = (pluginContext) => {
+
+  //  We use a loop here to provide user the ability to customize the interval.
+  //  it will keep looping unless got an exception.
+  start = (env = {}) => new Promise((resolve) => {
```
Should we return this directly instead of assigning to a variable? Or `const start`?

twang2218 replied:
the `start()` will be called later, so it cannot be anonymous.

I think I missed the `const` for all the functions, 😂, I will fix that.

blainesch commented:
Looks great. I'll merge after that last comment! [:
